src/python/pants/backend/python/target_types_rules.py

```diff
@@ -181,12 +185,38 @@ def _classify_entry_points(
 async def resolve_python_distribution_entry_points(
     request: ResolvePythonDistributionEntryPointsRequest,
 ) -> ResolvedPythonDistributionEntryPoints:
-    field_value = request.entry_points_field.value
-    if field_value is None:
+    if request.entry_points_field:
+        address = request.entry_points_field.address
+        dist_field_value = request.entry_points_field.value
+        if dist_field_value is not None:
+            all_entry_points = cast(_EntryPointsDictType, dist_field_value)
+        else:
+            return ResolvedPythonDistributionEntryPoints()
+
+    elif request.provides_field:
+        address = request.provides_field.address
+        provides_field_value = cast(
+            _EntryPointsDictType, request.provides_field.value.kwargs.get("entry_points") or {}
+        )
+
+        with_binaries = request.provides_field.value.binaries
+        if with_binaries:
+            # we want to be careful not to mutate the kwargs of the python artifact (setup_py)
+            all_entry_points = {
+                **provides_field_value,
+                "console_scripts": {
+                    **provides_field_value.get("console_scripts", {}),
+                    **with_binaries,
+                },
+            }
+        elif provides_field_value:
+            all_entry_points = provides_field_value
+        else:
+            return ResolvedPythonDistributionEntryPoints()
+    else:
         return ResolvedPythonDistributionEntryPoints()
 
-    address = request.entry_points_field.address
-    classified_entry_points = list(_classify_entry_points(field_value))
+    classified_entry_points = list(_classify_entry_points(all_entry_points))
```

Comment:
Had a bit of a struggle with MyPy here (not yet that experienced with python type hints..)
So ended up resolving to using a few intermediary variables to overcome the `Optional`s. If there is a better way....

Reply:
One way to do this is to set `all_entry_points: MyType[Foo, Bar]` at the top of this whole conditional block. Note that you don't set the actual value to anything. It's only to let MyPy know "hey I'm gonna have this variable with this type".

Reply:
Will try that next time. Not feeling like digging around any more in this as it is working, unless it's too painful for you eyes ;)